src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpWindowsProxy.cs

```diff
@@ -166,6 +179,21 @@ private HttpWindowsProxy(WinInetProxyHelper proxyHelper, SafeWinHttpHandle sessi
                     }
                 }
             }
+
+            _cachedProxies = new ConcurrentDictionary<string, CachedProxy>();
+            _cacheCleanupTimer = new Timer(state =>
+            {
+                var cachedProxies = (ConcurrentDictionary<string, CachedProxy>)state;
+                int ticks = Environment.TickCount;
+
+                foreach (KeyValuePair<string, CachedProxy> kvp in cachedProxies)
+                {
+                    if ((ticks - kvp.Value.LastAccessTicks) > CacheCleanupAgeMilliseconds)
+                    {
+                        cachedProxies.TryRemove(kvp.Key, out _);
+                    }
+                }
+            }, _cachedProxies, CacheCleanupFrequencyMilliseconds, CacheCleanupFrequencyMilliseconds);
```

Comment:
It's rare that we'll have multiple proxies, right? (Otherwise we would have hit this issue long ago?)  Assuming yes, it'd be good to avoid starting this recurring timer until it's actually required.

Reply:
It'd also be good to understand if this cache is really needed at all.  If, for example, our policy was to just always try each proxy in order, we wouldn't need the cache at all.  And I worry a bit that trying to optimize allocations for the case of multiple proxies will actually be penalizing the much more common case.

Reply:
> It's rare that we'll have multiple proxies, right? (Otherwise we would have hit this issue long ago?) Assuming yes, it'd be good to avoid starting this recurring timer until it's actually required.

An alternate, avoiding the timer, is to include a counter on `GetMultiProxy` and flush the cache every X counts. My assumption is that there will generally not be a large number of unique proxy strings, so this might be more appropriate.



> It'd also be good to understand if this cache is really needed at all. If, for example, our policy was to just always try each proxy in order, we wouldn't need the cache at all. And I worry a bit that trying to optimize allocations for the case of multiple proxies will actually be penalizing the much more common case.

I think a cache is needed to avoid severe performance issues of retrying failing proxies for every request. 
One other option I considered is to have a single "_cachedProxy" with the last string parsed, but this has the same issue for non-trivial PAC scripts.

Cache is not really there to optimize allocations; that is simply a happy side-effect. The common case of having only a single proxy isn't really hurt here -- it actually avoids parsing/allocating a new Uri for every request.

Reply:
Timer removed.